internal/pgkeyvalue/pgkeyvalue.go

```diff
@@ -0,0 +1,198 @@
+package pgkeyvalue
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/hashicorp/golang-lru/simplelru"
+	"github.com/jackc/pgconn"
+	"github.com/jackc/pgx/v4"
+	"github.com/jackc/pgx/v4/pgxpool"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+
+	"github.com/G-Research/armada/internal/common/armadaerrors"
+	"github.com/G-Research/armada/internal/common/logging"
+)
+
+// PGKeyValueStore is a time-limited key-value store backed by postgres with a local LRU cache.
+// The store is write-only, i.e., writing to an existing key will return an error (of type *armadaerrors.ErrAlreadyExists).
+// Keys can only be deleted by running the cleanup function.
+// Deleting keys does not cause caches to update, i.e., nodes may have an inconsistent view if keys are deleted.
+type PGKeyValueStore struct {
+	// For performance, keys are cached locally.
+	cache *simplelru.LRU
+	// Postgres connection.
+	db *pgxpool.Pool
+	// Name of the postgres table used for storage.
+	tableName string
+}
+
+func New(db *pgxpool.Pool, tableName string) (*PGKeyValueStore, error) {
+	if db == nil {
+		return nil, errors.WithStack(&armadaerrors.ErrInvalidArgument{
+			Name:    "db",
+			Value:   db,
+			Message: "db must be non-nil",
+		})
+	}
+	if tableName == "" {
+		return nil, errors.WithStack(&armadaerrors.ErrInvalidArgument{
+			Name:    "TableName",
+			Value:   tableName,
+			Message: "TableName must be non-empty",
+		})
+	}
+	cache, err := simplelru.NewLRU(100, nil)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	return &PGKeyValueStore{
+		cache:     cache,
+		db:        db,
+		tableName: tableName,
+	}, nil
+}
+
+// Add adds a key-value pair. Returns true if successful and false if the key already exists.
+// The posgres table backing the key-value storage is created automatically if it doesn't already exist.
+func (c *PGKeyValueStore) Add(ctx context.Context, key string, value []byte) (bool, error) {
+	ok, err := c.add(ctx, key, value)
+
+	// If the table doesn't exist, create it and try again.
+	var pgErr *pgconn.PgError
+	if errors.As(err, &pgErr) && pgErr.Code == "42P01" { // Relation doesn't exist; create it.
+		c.createTable(ctx)
+		ok, err = c.add(ctx, key, value)
+	}
+
+	return ok, err
+}
+
+// AddKey is equivalent to Add(ctx, key, nil).
+func (c *PGKeyValueStore) AddKey(ctx context.Context, key string) (bool, error) {
+	return c.Add(ctx, key, nil)
+}
+
+func (c *PGKeyValueStore) createTable(ctx context.Context) error {
+	var pgErr *pgconn.PgError
+	_, err := c.db.Exec(ctx, fmt.Sprintf("create table %s (key text primary key, value bytea, inserted timestamp not null);", c.tableName))
+	if errors.As(err, &pgErr) && pgErr.Code == "42P07" { // Relation already exists (i.e., someone else just created it), which is fine.
```

Comment:
I think pgx has some constants for errors

Reply:
Nice. I've updated the code to use them.